sqlair.go

```diff
@@ -55,31 +96,46 @@ func MustPrepare(query string, typeSamples ...any) *Statement {
 }
 
 type DB struct {
-	db *sql.DB
+	cacheID dbID
+	db      *sql.DB
 }
 
-func NewDB(db *sql.DB) *DB {
-	return &DB{db: db}
+func dbFinalizer(db *DB) {
+	// There is no need to close the sql.Stmts here, the resources in the
+	// database are freed on db.Close.
+	cacheMutex.Lock()
+	defer cacheMutex.Unlock()
+	db.db.Close()
+	stmtCache := dbStmts[db.cacheID]
+	for stmtCacheID, _ := range stmtCache {
+		delete(stmtDBCache[stmtCacheID], db.cacheID)
+	}
+	delete(dbStmts, db.cacheID)
+}
+
+func NewDB(sqldb *sql.DB) *DB {
+	cacheID := atomic.AddInt64(&dbIDCount, 1)
+	cacheMutex.Lock()
+	dbStmts[cacheID] = make(map[stmtID]bool)
+	cacheMutex.Unlock()
+	var db = DB{db: sqldb, cacheID: cacheID}
+	runtime.SetFinalizer(&db, dbFinalizer)
+	return &db
 }
 
 // PlainDB returns the underlying database object.
 func (db *DB) PlainDB() *sql.DB {
 	return db.db
 }
 
-// querySubstrate abstracts the different surfaces that the query can be run on.
-// For example, the database or a transaction.
-type querySubstrate interface {
-	QueryContext(ctx context.Context, sql string, args ...any) (*sql.Rows, error)
-	ExecContext(ctx context.Context, sql string, args ...any) (sql.Result, error)
-}
-
 // Query holds the results of a database query.
 type Query struct {
-	qe  *expr.QueryExpr
-	qs  querySubstrate
-	ctx context.Context
-	err error
+	qe   *expr.QueryExpr
+	stmt *sql.Stmt
```

Comment:
We need a way to differentiate between our stmt and the sql stmt.  How about using a convention that is actually a bit unusual in terms of casing: our variables are named `tx`, `db`, and `stmt`, while the ones in the sql package are named `sqltx`, `sqldb`, and `sqlstmt`. How does that sound?

If accepted, let's please have a pass over the diff in this PR to clean up the use cases so we flag each of these appropriately.

Reply:
I'll make this change